qutebrowser/misc/tabsfinder.py

```diff
@@ -0,0 +1,92 @@
+# vim: ft=python fileencoding=utf-8 sts=4 sw=4 et:
+
+import re
+from collections import OrderedDict, namedtuple
+
+from qutebrowser.utils import objreg
+
+Tab = namedtuple('Tab', 'url title index')
+
+
+class TabsFinder:
+    """
+    TabsFinder is in charge of gathering a reusable list of
+    tabs amongst all windows
+
+    Args:
+        tabbed_browser:
+        skip_win_id: The id of the window to skip, or None to include all.
+    """
+
+    def __init__(self, skip_win_id=None):
+        self._skip_win_id = skip_win_id
+        self._tabs = []
+        self._tabs_by_win_id = OrderedDict()
+
+        self._find()
+
+    def _find(self):
+        """Find all the tabs
+
+        """
+
+        for win_id in objreg.window_registry:
+            tabbed_browser = objreg.get('tabbed-browser', scope='window', window=win_id)
+            self._tabs_by_win_id[win_id] = []
+
+            if self._skip_win_id is not None and win_id == self._skip_win_id:
+                continue
+
+            if tabbed_browser.shutting_down:
+                continue
+
+            for idx in range(tabbed_browser.widget.count()):
+                tab = tabbed_browser.widget.widget(idx)
+                t = Tab(tab.url().toDisplayString(),
+                        tabbed_browser.widget.page_title(idx),
+                        "{}/{}".format(win_id, idx + 1))
+
+                self._tabs.append(t)
+                self._tabs_by_win_id[win_id].append(t)
+
+    def tabs_by_window(self):
+        """
+        Returns an iteratable set of items from win_id to tabs
+        """
+
+        return self._tabs_by_win_id.items()
+
+    def tabs(self):
+        return self._tabs
+
+    def index_for_url(self, url):
+        for tab in self._tabs:
+            if url == tab.url:
+                return tab.index
+
+        return None
+
+
+def delete_tab(data):
+    """Close the selected tab."""
+
+    pattern = "^\d+/\d+$"
+
+    if type(data) is str:
+        index = data
+    elif type(data) is Tab:
```

Comment:
This is only ever called from a completion category, which means the type should always be `list`, right?
Why the switches for other data types?

Functions that operate on a wide array of types are tricky, and I _think_ there's a movement towards adding typing to parts of qutebrowser now?